library/init/data/nat/bitwise.lean

```diff
@@ -11,15 +11,78 @@ universe u
 
 namespace nat
 
+  def bodd_div2 : ℕ → bool × ℕ
+  | 0        := (ff, 0)
+  | (succ n) :=
+    match bodd_div2 n with
+    | (ff, m) := (tt, m)
+    | (tt, m) := (ff, succ m)
+    end
+
+  def div2 (n : ℕ) : ℕ := (bodd_div2 n).2
+
+  def bodd (n : ℕ) : bool := (bodd_div2 n).1
```

Comment:
Oooh, I see.  Yes, this is really, really ugly.  And we will probably see this again in the future, since it happens with every WF definition...

I see two potential solutions:
1. Use equational rfl-lemmas for reduction in the type_context.  This is a bit of work, and Leo has mentioned that there are lots of issues here.
2. As a quick hack, we could special-case `acc.rec` (and maybe `eq.rec` and `heq.rec`?) in the type_context by reducing the major premise using `transparency_mode::All`.

BTW, we have exactly the same issue with `attribute [irreducible]`.